R/pivot-long.R

```diff
@@ -233,30 +233,28 @@ pivot_longer_spec <- function(data,
     val_type <- vec_ptype_common(!!!set_names(val_cols[col_id], cols), .ptype = values_ptypes[[value]])
     out <- vec_c(!!!val_cols, .ptype = val_type)
     # Interleave into correct order
-    idx <- (matrix(seq_len(nrow(data) * length(val_cols)), ncol = nrow(data), byrow = TRUE))
+    # TODO somehow `t(matrix(x))` is _faster_ than `matrix(x, byrow = TRUE)`
```

Comment:
Do you think it's worth pulling this out into an `interleave` function? I'm not sure this comes up very often, but we could also move into vctrs and consider our own C implementation (which I think should be fairly straight forward)

Reply:
I also think it's not that common but it would be a great help here to have a better version. The current version is quite memory and time consuming

![image](https://user-images.githubusercontent.com/18258854/122526742-ebe03a80-d01a-11eb-9626-add7c62c63ff.png)

By the way do you have any idea if there would be any performance benefit if instead of using `vec_slice()` to reorder a vector like in this case there were a function `vec_reorder()` that only reorders the elements of it? I could imagine `vec_reorder()` to be more efficient but I don't know much about C code and the vctrs internals.

Reply:
And do you think it makes sense to create a bug ticket for R that `t(matrix(x))` is faster than `matrix(x, byrow = TRUE)`? Because this does look like a bug to me.

Reply:
You mean for the constraint that there's a 1-to-1 mapping of elements of the input and output vectors? I don't see any obvious performance implication of specialising (there's probably some minor improvements to be had if you could skip some error checking, but that's often not worth it in the long term).

You can try filing an bug with R-core, but the process is a bit painful, and it's often hard to convince them that a problem exists.

Reply:
> You mean for the constraint that there's a 1-to-1 mapping of elements of the input and output vectors? I don't see any obvious performance implication of specialising (there's probably some minor improvements to be had if you could skip some error checking, but that's often not worth it in the long term).

Yes, that's what I meant.

> You can try filing an bug with R-core, but the process is a bit painful, and it's often hard to convince them that a problem exists.

That's what I was kind of expected to hear, so I'd rather not do that (at least for now).